   # Lab 6 - Solution

The template is given as a starting point to help you get up to speed with loading file and understanding data.  
Use the template and answer each question in the appropriate cell.

Submit:
- eLearning

Naming: First-name_Last_name_ID_Lab_6.ipynb

## Dataset description

The dataset for Lab-6 is sample of the SBA dataset posted on Kaggle. 

The dataset is from the U.S. Small Business Administration (SBA)
The U.S. SBA was founded in 1953 on the principle of promoting and assisting small enterprises in the U.S. credit market 
(SBA Overview and History, US Small Business Administration (2015)). Small businesses have been a primary source of job creation in the United States; therefore, fostering small business formation and growth has social benefits by creating job opportunities and reducing unemployment.
There have been many success stories of start-ups receiving SBA loan guarantees such as FedEx and Apple Computer. 
However, there have also been stories of small businesses and/or start-ups that have defaulted on their SBA-guaranteed loans.

More info on the original dataset: https://www.kaggle.com/mirbektoktogaraev/should-this-loan-be-approved-or-denied


## Question 1

Load same data into H2O-3 dataframe.  
Describe imported dataframe.

In [1]:
import h2o
try:
    h2o.init()
    h2o.cluster().shutdown()
except:
    pass 

Checking whether there is an H2O instance running at http://localhost:54321 . connected.


--------------------------  --------------------------------
H2O_cluster_uptime:         5 days 21 hours 58 mins
H2O_cluster_timezone:       America/Chicago
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.38.0.4
H2O_cluster_version_age:    2 months and 7 days
H2O_cluster_name:           H2O_from_python_usmashnov_3rwo5k
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    3.993 Gb
H2O_cluster_total_cores:    12
H2O_cluster_allowed_cores:  12
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://localhost:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.8.10 final
--------------------------  --------------------------------

H2O session _sid_958c closed.


In [2]:
#Limit to 8 threads and 8GB memory
h2o.init(nthreads=8, max_mem_size=8)

Checking whether there is an H2O instance running at http://localhost:54321 ..... not found.
Attempting to start a local H2O server...
  Java Version: openjdk version "9"; OpenJDK Runtime Environment (build 9+181); OpenJDK 64-Bit Server VM (build 9+181, mixed mode)
  Starting server from /Users/usmashnov/var/virtual-env/ml-spring-2023/lib/python3.8/site-packages/h2o/backend/bin/h2o.jar
  Ice root: /var/folders/_z/jf3ghwdx1kg905xm5p1nktlh0000gp/T/tmpg6yhry7z
  JVM stdout: /var/folders/_z/jf3ghwdx1kg905xm5p1nktlh0000gp/T/tmpg6yhry7z/h2o_usmashnov_started_from_python.out
  JVM stderr: /var/folders/_z/jf3ghwdx1kg905xm5p1nktlh0000gp/T/tmpg6yhry7z/h2o_usmashnov_started_from_python.err
  Server is running at http://127.0.0.1:54321
Connecting to H2O server at http://127.0.0.1:54321 ... successful.


--------------------------  --------------------------------
H2O_cluster_uptime:         03 secs
H2O_cluster_timezone:       America/Chicago
H2O_data_parsing_timezone:  UTC
H2O_cluster_version:        3.38.0.4
H2O_cluster_version_age:    2 months and 7 days
H2O_cluster_name:           H2O_from_python_usmashnov_kxbrhs
H2O_cluster_total_nodes:    1
H2O_cluster_free_memory:    8 Gb
H2O_cluster_total_cores:    12
H2O_cluster_allowed_cores:  8
H2O_cluster_status:         locked, healthy
H2O_connection_url:         http://127.0.0.1:54321
H2O_connection_proxy:       {"http": null, "https": null}
H2O_internal_security:      False
Python_version:             3.8.10 final
--------------------------  --------------------------------

In [3]:
df_h = h2o.import_file('./data/SBA_loans_lab_6.zip')

Parse progress: |████████████████████████████████████████████████████████████████| (done) 100%


In [4]:
df_h.describe()

Rows:179833
Cols:20

City            State    Zip                 Bank                            BankState    NAICS               Term                NoEmp               NewExist             CreateJob           RetainedJob         FranchiseCode      UrbanRural          RevLineCr    LowDoc    DisbursementGross    BalanceGross        GrAppv             SBA_Appv            Defaulted
-------  --------------  -------  ------------------  ------------------------------  -----------  ------------------  ------------------  ------------------  -------------------  ------------------  ------------------  -----------------  ------------------  -----------  --------  -------------------  ------------------  -----------------  ------------------  -------------------
type     enum            enum     int                 enum                            enum         int                 int                 int                 int                  int                 int                 int                int                 enum         enum      int                  int                 int                int                 int
mins                              0.0                                                              0.0                 0.0                 0.0                 0.0                  0.0                 0.0                 0.0                0.0                                        0.0                  0.0                 500.0              250.0               0.0
mean                              53773.70054439398                                                399594.72817002435  110.90885988667262  11.384150851067382  1.28071853623269     9.011527361496505   11.433813593723066  2755.589947340032  0.7598827801349026                         201926.888624446     6.553808255436987   193398.5314152575  150058.3506586667   0.17522924046198418
maxs                              99999.0                                                          928120.0            461.0               8500.0              2.0                  8800.0              8800.0              99999.0            2.0                                        11446325.0           996262.0            5000000.0          4869000.0           1.0
sigma                             31182.082734237705                                               263331.7973921778   78.95806965689958   67.30104824719153   0.45206614960651104  246.60707495715346  247.46697982527493  12754.79777368782  0.646773081103855                          290056.8000335533    2367.8805220862832  284469.6765889608  229388.82282767945  0.380164119046022
zeros                             43                                                               40267               161                 1369                220                  125596              87605               42030              64388                                      44                   179829              0                  0                   148321
missing  3               4        0                   316                             318          0                   0                   0                   23                   0                   0                   0                  0                   932          855       0                    0                   0                  0                   0
0        Cuyahoga Falls  OH       44224.0             PNC BANK, NATIONAL ASSOCIATION  OH           332710.0            84.0                3.0                 1.0                  0.0                 3.0                 1.0                1.0                 N            N         99000.0              0.0                 99000.0            49500.0             0.0
1        LAWRENCEVILLE   GA       30043.0             MERRILL LYNCH BANK USA          UT           454390.0            84.0                1.0                 2.0                  0.0                 1.0                 1.0                1.0                 0            N         100900.0      

### Set attributes for the categorical  variables - make non-numerical columns "factor" (categorical)

We will be using H2O-3 DRF model. The model can handle categorical variable and missing values.  
We will need to **mark** categorical variables with `asfactor` method

In [5]:
# Choose which columns to encode
cat_columns = ["City","State","Bank","BankState", "NewExist", "RevLineCr","LowDoc","Zip"]
encoded_columns = cat_columns
response = "Defaulted"

df_h[encoded_columns+[response]] = df_h[encoded_columns+[response]].asfactor()

## Split to train/test/validation

In [6]:
train,test,valid = df_h.split_frame(ratios=[.7, .15], seed=123)

## Question 1 

Use H2O-3 to train Distributed Random Forest model with default parameters.  
Documentation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/drf.html#defining-a-drf-model

In [7]:
from h2o.estimators import H2ORandomForestEstimator

predictors = train.columns
predictors.remove("Defaulted")
print("Predictor columns:", predictors)    
response_col = "Defaulted"

drf = H2ORandomForestEstimator(nfolds=0,seed=123)

drf.train(predictors, response_col, training_frame= train)
model_summary = drf.summary()

Predictor columns: ['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv']
drf Model Build progress: |██████████████████████████████████████████████████████| (done) 100%


In [8]:
model_summary

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    50                 50                          5.29338e+06            20           20           20            5725          7517          6730.1

### Model performance metrics on Test dataset

Alternative method to calculate metrics is to use `h2o.make_metrics(pred, actual)` function:   
https://docs.h2o.ai/h2o/latest-stable/h2o-docs/performance-and-prediction.html#computing-model-metrics-from-general-predictions

In [9]:
print("Best accuracy threshold:",drf.model_performance(test).accuracy()[0][0],"\n",
      " Accuracy:",
      drf.model_performance(test).accuracy()[0][1])
print("Best F1 threshold:",drf.model_performance(test).F1()[0][0],"\n",
      " F1:",
      drf.model_performance(test).F1()[0][1])
print("Model AUC:", drf.model_performance(test).auc())
print("Model AUCPR:", drf.model_performance(test).aucpr())
drf.model_performance(test).confusion_matrix()

Best accuracy threshold: 0.456615068949759 
  Accuracy: 0.9346093691815
Best F1 threshold: 0.34488910898743647 
  F1: 0.8054782794778516
Model AUC: 0.9561638211354876
Model AUCPR: 0.8733678806650562


Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.34488910898743647
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21272  871   0.0393   (871.0/22143.0)
1      947    3764  0.201    (947.0/4711.0)
Total  22219  4635  0.0677   (1818.0/26854.0)

## Question 2

You are asked to tune the model from question #1. You should review full parameter list for the DRF model documentation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/drf.html#defining-a-drf-model

Pay special attention to the following parameters:
- balance_classes
- ntrees
- max_depth
- early stopping params: stopping_rounds, stopping_metric, stopping_tolerance
- mtries
- sample_rate

Alternative to find set `ntrees` parameter is to set it to high number, and use early stopping to find optimal number of trees. (We will do it in the next question)

Find best parameters using H2O GridSearch for the following search space.
Modify the example code below as needed to make it work.  
You can experiment with additional options/parameters to see if you can get a better model.
```
# DRF hyperparameters
from h2o.grid.grid_search import H2OGridSearch

# DRF hyperparameters
drf_params2  = {'ntrees': [25,50,100],
               'max_depth': [10,15,20,25]}

# Train and validate a cartesian grid of DRFs
drf_grid2  = H2OGridSearch(model=H2ORandomForestEstimator,
                          grid_id='drf_grid2',
                          hyper_params=drf_params2)

drf_grid2.train(x=predictors,
               y=response_col,
               nfolds=0,
               training_frame=train,
               validation_frame=valid,
               seed=123)
```

You can follow the progress of the Grid search job by using H2O Flow, the URL is shown right after `h2o.init()`.
Answer following questions:
- What is the best set of parameters for the DRF model using AUCPR metric?
- What is the best model performance (AUCPR) on a validation dataset?
- What is the best model performance (AUCPR) on a test dataset?
- Produce Confusion Matrix using best F1 probability threshold

In [10]:
from h2o.grid.grid_search import H2OGridSearch

predictors = train.columns
predictors.remove("Defaulted")
print("Predictor columns:", predictors)    
response_col = "Defaulted"


# DRF hyperparameters
drf_params2  = {'ntrees': [25,50,100],
               'max_depth': [10,15,20,25]}

# Train and validate a cartesian grid of DRFs
drf_grid2  = H2OGridSearch(model=H2ORandomForestEstimator,
                          grid_id='drf_grid2',
                          hyper_params=drf_params2)

drf_grid2.train(x=predictors, 
               y=response_col,
               nfolds=0,
               training_frame=train,
               validation_frame=valid,
               seed=123)

Predictor columns: ['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv']
drf Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%


Hyper-Parameter Search Summary: ordered by increasing logloss
    max_depth    ntrees    model_ids           logloss
--  -----------  --------  ------------------  ---------
    20           100       drf_grid2_model_11  0.185148
    25           100       drf_grid2_model_12  0.186262
    20           50        drf_grid2_model_7   0.188248
    15           100       drf_grid2_model_10  0.195359
    20           25        drf_grid2_model_3   0.195886
    15           50        drf_grid2_model_6   0.197195
    15           25        drf_grid2_model_2   0.198637
    25           50        drf_grid2_model_8   0.203982
    10           100       drf_grid2_model_9   0.234752
    10           50        drf_grid2_model_5   0.235868
    10           25        drf_grid2_model_1   0.236234
    25           25        drf_grid2_model_4   0.242712

In [11]:
# Get the grid results, sorted by validation AUCPR
drf_gridperf2 = drf_grid2.get_grid(sort_by='aucpr', decreasing=True)
drf_gridperf2

Hyper-Parameter Search Summary: ordered by decreasing aucpr
    max_depth    ntrees    model_ids           aucpr
--  -----------  --------  ------------------  --------
    25           100       drf_grid2_model_12  0.880707
    20           100       drf_grid2_model_11  0.879914
    25           50        drf_grid2_model_8   0.875909
    20           50        drf_grid2_model_7   0.875636
    15           100       drf_grid2_model_10  0.871812
    25           25        drf_grid2_model_4   0.870613
    20           25        drf_grid2_model_3   0.869909
    15           50        drf_grid2_model_6   0.868217
    15           25        drf_grid2_model_2   0.863939
    10           100       drf_grid2_model_9   0.831844
    10           50        drf_grid2_model_5   0.829329
    10           25        drf_grid2_model_1   0.824572

In [12]:
# Grab the top DRF model, chosen by validation AUCPR
best_drf2 = drf_gridperf2.models[0]

# Now let's evaluate the model performance on a Validation set
# We should see same metric as in the grid
best_drf_perf2 = best_drf2.model_performance(valid=True)
best_drf_perf2.aucpr()

0.8807072153213351

In [13]:
# Now let's evaluate the model performance on a test set
# so we get an honest estimate of top model performance
best_drf_perf2 = best_drf2.model_performance(test_data=test)
best_drf_perf2.aucpr()

0.8794417755403234

In [14]:
best_drf2.confusion_matrix(valid=True)

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.3928236840479076
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21556  662   0.0298   (662.0/22218.0)
1      1011   3781  0.211    (1011.0/4792.0)
Total  22567  4443  0.0619   (1673.0/27010.0)

In [15]:
# Example of using different threshold

best_drf2.confusion_matrix(valid=True, thresholds=0.5)

Could not find exact threshold 0.5; using closest threshold found 0.5001184068520864.


Confusion Matrix (Act/Pred) @ threshold = 0.5001184068520864
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21815  403   0.0181   (403.0/22218.0)
1      1325   3467  0.2765   (1325.0/4792.0)
Total  23140  3870  0.064    (1728.0/27010.0)

## Question 3

You are asked to tune the model from question #1. You should review full parameter list for the DRF model documentation: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/data-science/drf.html#defining-a-drf-model

Tune model using early stopping criteria by specifying `ntrees`=1000 and setting the parameters required for early stopping. You will need to provide Validation data-frame for early stopping to work correctly.


Alternative to find optimal setting for `ntrees` (number of trees in RF) parameter is to set it to high number, and use early stopping to find optimal number of trees.

Find best parameters using H2O GridSearch for the following search space.
You should experiment with additional options/parameters to see if you can train better models.
The goal is to get a better model than in Q2. Add at least one additional hyper-parameter to tune.
Minimum parameters to tune:
```
drf_params  = {'max_depth': [10,15, 25]}
```
You should review early stopping parameters to be added to the `train` method:
- stopping_rounds
- stopping_metric
- stopping_tolerance

               
Answer following questions:
- What is the best set of parameters for the DRF model using AUCPR metric?
- What is the best model performance (AUCPR) on a validation dataset?
- What is the best model performance (AUCPR) on a test dataset?
- Produce Confusion Matrix using best F1 probability threshold

In [16]:
from h2o.grid.grid_search import H2OGridSearch

# DRF hyperparameters
drf_params3  = {'max_depth': [10,15,25],
                'stopping_rounds':[5, 8]}

# Train and validate a cartesian grid of DRFs
drf_grid3  = H2OGridSearch(model=H2ORandomForestEstimator,
                          grid_id='drf_grid3',
                          hyper_params=drf_params3)

drf_grid3.train(x=predictors, 
                y=response_col,
                nfolds=0,
                score_tree_interval=5,
                stopping_metric='AUCPR',
                ntrees=1000,
                training_frame=train,
                validation_frame=valid,
                seed=123)

drf Grid Build progress: |███████████████████████████████████████████████████████| (done) 100%


Hyper-Parameter Search Summary: ordered by increasing logloss
    max_depth    stopping_rounds    model_ids          logloss
--  -----------  -----------------  -----------------  ---------
    25           8                  drf_grid3_model_6  0.183289
    25           5                  drf_grid3_model_3  0.186262
    15           8                  drf_grid3_model_5  0.195194
    15           5                  drf_grid3_model_2  0.195359
    10           8                  drf_grid3_model_4  0.234271
    10           5                  drf_grid3_model_1  0.234752

In [17]:
# Get the grid results, sorted by validation AUCPR
drf_gridperf3 = drf_grid3.get_grid(sort_by='aucpr', decreasing=True)
drf_gridperf3

Hyper-Parameter Search Summary: ordered by decreasing aucpr
    max_depth    stopping_rounds    model_ids          aucpr
--  -----------  -----------------  -----------------  --------
    25           8                  drf_grid3_model_6  0.881277
    25           5                  drf_grid3_model_3  0.880707
    15           8                  drf_grid3_model_5  0.872433
    15           5                  drf_grid3_model_2  0.871812
    10           8                  drf_grid3_model_4  0.832768
    10           5                  drf_grid3_model_1  0.831844

In [18]:
# Grab the top DRF model, chosen by validation AUCPR
best_drf3 = drf_gridperf3.models[0]

# Now let's evaluate the model performance on a Validation set
# We should see same metric as in the grid
best_drf_perf3 = best_drf3.model_performance(valid)
best_drf_perf3.aucpr()

0.881276608934421

In [19]:
# Now let's evaluate the model performance on a test set
# so we get an honest estimate of top model performance
best_drf_perf3 = best_drf3.model_performance(test_data=test)
best_drf_perf3.aucpr()

0.8802341215844708

In [20]:
best_drf3.confusion_matrix(valid=True)

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.3837758579912285
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21510  708   0.0319   (708.0/22218.0)
1      978    3814  0.2041   (978.0/4792.0)
Total  22488  4522  0.0624   (1686.0/27010.0)

In [21]:
best_drf3.model_performance(test_data=test).confusion_matrix()

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.35945941268744175
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21342  801   0.0362   (801.0/22143.0)
1      947    3764  0.201    (947.0/4711.0)
Total  22289  4565  0.0651   (1748.0/26854.0)

#### Find number of trees for the best model

In [22]:
best_drf3.summary()

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    150                150                         1.88617e+07            25           25           25            6803          9747          8044.24

## Question 4

Use best model parameters from the GridSearch in question #3. Train new model as following:
- Train model on a combined dataset: train+validation
- Don't use cross-validation - `nfolds=0`
- Evaluate trained model on test dataset
- Compare AUCPR (test dataset) metrics from the question #2 and #3

In [23]:
from h2o.grid.grid_search import H2OGridSearch

predictors = train.columns
predictors.remove("Defaulted")
print("Predictor columns:", predictors)    
response_col = "Defaulted"



train_val = train.rbind(valid)
drf_model_full = H2ORandomForestEstimator(max_depth=25,
                                          ntrees=150,                              
                                          nfolds=0,
                                          seed=123)
drf_model_full.train(predictors, response_col, training_frame= train_val)
model_summary = drf_model_full.summary()

Predictor columns: ['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv']
drf Model Build progress: |██████████████████████████████████████████████████████| (done) 100%


In [24]:
model_summary

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    150                150                         2.24694e+07            25           25           25            8107          11003         9574.23

In [25]:
print("Model AUCPR on Testing dataset:", drf_model_full.model_performance(test).aucpr())

Model AUCPR on Testing dataset: 0.8842963216215659


Model trained on train dataset only produced AUCPR:          0.8802341215844708    
Model trained on train+validation dataset produced AUCPR:    0.8842963216215659

The AUCPR of the model that used more data (train+valid) is slightly better.  

### Function to score the dataset and calculate confusion matrix

In [26]:
def conf_matrix(model,df_h2o,target_col, threshold):
    from sklearn.metrics import confusion_matrix
    import numpy as np
    
    pred_df = model.predict(df_h2o)
    pred_df = pred_df.cbind(df_h2o)[[target_col,"p1"]]
    pred_df_pd = pred_df.as_data_frame()
    pred_df_pd["predicted_class"] = np.where(pred_df_pd['p1'] >threshold, 1, 0)
    cm = confusion_matrix(pred_df_pd[target_col], pred_df_pd["predicted_class"])
    return cm

In [27]:
drf_model_full.model_performance(test_data=test).confusion_matrix()

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.35942694245378176
       0      1     Error    Rate
-----  -----  ----  -------  ----------------
0      21356  787   0.0355   (787.0/22143.0)
1      937    3774  0.1989   (937.0/4711.0)
Total  22293  4561  0.0642   (1724.0/26854.0)

In [28]:
# Set probability threshold to determine class
threshold=0.35942694245378176
threshold = drf_model_full.model_performance(test_data=test).F1()[0][0]

conf_matrix(model=drf_model_full, 
            df_h2o=test, 
            target_col=response_col, 
            threshold=threshold)

drf prediction progress: |███████████████████████████████████████████████████████| (done) 100%


array([[21358,   785],
       [  947,  3764]])

The difference is because of floating point approximation between H2O and Pandas dataframes.

# Question 5

Review following documentation section first: https://docs.h2o.ai/h2o/latest-stable/h2o-docs/cross-validation.html

- Train DRF model with the parameters found in the Question 3.
- Use full data to train the model.
- Use cross-validation to produce reliable metrics for the model performance
- Produce model metrics using cross-validation 

**Note**: understand why the model training takes significantly longer than before we used cross-validation

In [29]:
predictors = train.columns
predictors.remove("Defaulted")
print("Predictor columns:", predictors)    
response_col = "Defaulted"

drf_model_xval = H2ORandomForestEstimator(max_depth=25,
                                          ntrees=150,                              
                                          nfolds=5,
                                          seed=123)
drf_model_xval.train(predictors, response_col, training_frame= df_h)
model_summary = drf_model_xval.summary()

Predictor columns: ['City', 'State', 'Zip', 'Bank', 'BankState', 'NAICS', 'Term', 'NoEmp', 'NewExist', 'CreateJob', 'RetainedJob', 'FranchiseCode', 'UrbanRural', 'RevLineCr', 'LowDoc', 'DisbursementGross', 'BalanceGross', 'GrAppv', 'SBA_Appv']
drf Model Build progress: |██████████████████████████████████████████████████████| (done) 100%


In [30]:
model_summary

Model Summary: 
    number_of_trees    number_of_internal_trees    model_size_in_bytes    min_depth    max_depth    mean_depth    min_leaves    max_leaves    mean_leaves
--  -----------------  --------------------------  ---------------------  -----------  -----------  ------------  ------------  ------------  -------------
    150                150                         2.60307e+07            25           25           25            9335          12417         11055.6

### Correct way to produce model performance metrics using Hold-out data

In [31]:
print("Model AUCPR on Cross-Validation (hold-out) folds:", drf_model_xval.aucpr(xval=True))

Model AUCPR on Cross-Validation (hold-out) folds: 0.8807464701512521


In [32]:
drf_model_xval.model_performance(xval=True).confusion_matrix()

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.37356166234920907
       0       1      Error    Rate
-----  ------  -----  -------  ------------------
0      143365  4956   0.0334   (4956.0/148321.0)
1      6395    25117  0.2029   (6395.0/31512.0)
Total  149760  30073  0.0631   (11351.0/179833.0)

### Incorrect way to produce model metrics

In the below example we use training dataset to produce the metrics.
This is incorrect because the model will always overfit to train dataset and the metrics will not be reliable.

In [33]:
drf_model_xval.model_performance(df_h).confusion_matrix()

Confusion Matrix (Act/Pred) for max f1 @ threshold = 0.472630156462914
       0       1      Error    Rate
-----  ------  -----  -------  ----------------
0      148233  88     0.0006   (88.0/148321.0)
1      166     31346  0.0053   (166.0/31512.0)
Total  148399  31434  0.0014   (254.0/179833.0)

Calculate Confusion matrix manually, by scoring the data and using threshold that optimizes F1.

In [34]:
drf_model_xval.model_performance(test_data=df_h).F1()[0][0]

0.472630156462914

In [35]:
# Set probability threshold to determine class
threshold=0.472630156462914

conf_matrix(model=drf_model_xval, 
            df_h2o=df_h, 
            target_col=response_col, 
            threshold=threshold)

drf prediction progress: |███████████████████████████████████████████████████████| (done) 100%


array([[148234,     87],
       [   172,  31340]])